### PR TITLE
Rename and reorg types

### LINKED ISSUE #457: Rename and reorganise types a bit
Currently (v0.3) we have many types prefixed with `JsonRpc`, e.g. `JsonRpcRequest`, `JsonRpcInvalidRequest`, `JsonRpcResponse`, `JsonRpcSubscriptionParams` etc.

Rename such types without the `JsonRpc` prefix, `Request`, `InvalidRequest`, `Response` etc.

The rationale for adding the `JsonRpc` prefix in the first place was to signal that the type is part of [the spec](). It'd be good to continue having a clear separation between types stemming from the spec and types that are implementation specific. Today, most types in `jsonrpsee-types::v2` are from the spec.

While rummaging with the types it might be a good time to also figure out what to do about: 
- `v2::params::JsonRpcSubscriptionParams`: maybe not actually parameters? I think it's the subscription item response type?
- `v2::request::JsonRpcNotification`: isn't this a response types, and if yes, let's move it to `v2::response` (EDIT: see [the conversation here](https://github.com/paritytech/jsonrpsee/pull/459#discussion_r706199657))
- https://github.com/paritytech/jsonrpsee/blob/e62a0decd7cad25b7aea72e4e783d8322e99f7f6/types/src/v2/request.rs#L149

### PR DESCRIPTION
**Pull request by dvdplm**
⚠️  Breaking change ⚠️ 

Closes #457﻿.

The changes here mostly removes the `JsonRpc` prefix from the types that come from the spec. I have kept a prefix for some types with a very common stdlib version such as `Result` and `Error`. I also changed the serializable versions to consistently use the `*Ser` postfix (not loving this convention but have no better suggestion).

I'm tempted to make the `jsonrpsee-types` paths more shallow, i.e. re-export all the modules under `v2`, so that instead of doing `jsonrpsee-types::v2::error::RpcError` we'd have `jsonrpsee-types::v2::RpcError`; instead of `jsonrpsee-types::v2::response::Response` we'd have `jsonrpsee-types::v2::Response` etc. Thoughts?

### CODE REVIEW

**Comment by maciejhirsz** [comment]
How about `CallParams` instead of `ParamsSer`? Or, if we want to be slightly confusing but pedantically correct: `Arguments`.

**Comment by dvdplm** [comment]
> How about `CallParams` instead of `ParamsSer`? Or, if we want to be slightly confusing but pedantically correct: `Arguments`.

GH threw away my reply to this, grr.

I like `Arguments` quite a lot, but I don't think it's wise to move away from "*params*" because the spec uses "parameter" throughout and I don't think it's good to ask users/readers to map from one to the other.

The downside of going with `CallParams` is that it doesn't help with the other `*Ser` types: `RequestSer` and `NotificationSer`. Using a post-fix has the advantage of clearly communicating what is peculiar about all three of them.

:/

**Comment by niklasad1** [comment]
> I'm tempted to make the jsonrpsee-types paths more shallow, i.e. re-export all the modules under v2, so that instead of doing jsonrpsee-types::v2::error::RpcError we'd have jsonrpsee-types::v2::RpcError; instead of jsonrpsee-types::v2::response::Response we'd have jsonrpsee-types::v2::Response etc. Thoughts?

Sounds good to me

File: utils/src/server/rpc_module.rs [hunk 1]
```diff
@@ -780,16 +759,15 @@ mod tests {
 		let (sub_id, mut my_sub_stream) = module.test_subscription("my_sub", None).await;
 		for i in (0..=2).rev() {
 			let my_sub = my_sub_stream.next().await.unwrap();
-			let my_sub = serde_json::from_str::<JsonRpcNotification<JsonRpcSubscriptionParams<char>>>(&my_sub).unwrap();
+			let my_sub = serde_json::from_str::<Notification<SubscriptionPayload<char>>>(&my_sub).unwrap();
```
Comment (niklasad1) [inline utils/src/server/rpc_module.rs:762]: can't you use the `SubscriptionResponse` type alias here instead?

File: utils/src/server/rpc_module.rs [hunk 2]
```diff
@@ -780,16 +759,15 @@ mod tests {
...
 			assert_eq!(my_sub.params.result, std::char::from_digit(i, 10).unwrap());
 			assert_eq!(my_sub.params.subscription, v2::params::SubscriptionId::Num(sub_id));
 		}
 
 		// The subscription is now closed
 		let my_sub = my_sub_stream.next().await.unwrap();
 		let my_sub =
-			serde_json::from_str::<JsonRpcNotification<JsonRpcSubscriptionParams<SubscriptionClosedError>>>(&my_sub)
-				.unwrap();
+			serde_json::from_str::<Notification<SubscriptionPayload<SubscriptionClosedError>>>(&my_sub).unwrap();
```
Comment (niklasad1) [inline utils/src/server/rpc_module.rs:770]: can't you use the SubscriptionResponse type alias here instead?

File: ws-client/src/helpers.rs [hunk 3]
```diff
@@ -73,19 +73,19 @@ pub fn process_batch_response(manager: &mut RequestManager, rps: Vec<JsonRpcResp
 ///
 /// Returns `Ok()` if the response was successfully sent to the frontend.
 /// Return `Err(None)` if the subscription was not found.
-/// Returns `Err(Some(msg))` if the subscription was full.
+/// Returns `Err(Some(msg))` if the channel to the frontend was full.
```
Comment (niklasad1) [inline ws-client/src/helpers.rs:76]: I don't mind this much but I would call the `subscription stream` as a "frontend" as it's part of the `Subscription` returned to user after a subscription has been established and "not a frontend" in the client....

File: utils/src/server/rpc_module.rs [hunk 4]
```diff
@@ -506,7 +487,7 @@ impl<Context: Send + Sync + 'static> RpcModule<Context> {
 		Ok(())
 	}
 
-	/// Register an `alias` name for an `existing_method`.
+	/// Register an alias for an existing_method. Alias uniqueness is enforced at compile time.
```
Comment (niklasad1) [inline utils/src/server/rpc_module.rs:490]: This is only true for the proc macros and not in this context but for most use-cases your comment is true :P

```suggestion
	/// Register an alias for an existing_method.
```

**Comment by niklasad1** [comment]
> How about CallParams instead of ParamsSer? Or, if we want to be slightly confusing but pedantically correct: Arguments.

We have used `Params` because it is was spec states. I think `CallParams` is slightly confusing because when the server receives the params it's still a call?! At least `ParamsSer` is super obvious too me even if it's a bit "verbose"

`Arguments` could work though....

**Review by niklasad1 — APPROVED** [review]
LGTM, some minor comments to address

In an followup PR, we could flatten the `v2 mod` and remove the child modules `request, response, params`.

File: utils/src/server/rpc_module.rs (hunk 4 above)
Comment (dvdplm) [reply utils/src/server/rpc_module.rs:490]: Would `Alias uniqueness is enforced.` be correct?

**Comment by dvdplm** [comment]
> In an followup PR, we could flatten the `v2 mod` and remove the child modules `request, response, params`.

https://github.com/paritytech/jsonrpsee/issues/468

File: utils/src/server/rpc_module.rs (hunk 1 above)
Comment (dvdplm) [reply utils/src/server/rpc_module.rs:762]: Ofc I can, that's the very reason I added `SubscriptionResponse`! I messed up a merge I think. Ty!

File: utils/src/server/rpc_module.rs (hunk 4 above)
Comment (niklasad1) [reply utils/src/server/rpc_module.rs:490]: Yes, a duplicated alias will return error

**Review by maciejhirsz — APPROVED** [review]
Aside form my initial bikeshed it looks good to me :)

**Pull request merged**